### PR TITLE
Migrate JobIntentService to WorkManager

### PR DESCRIPTION
**Pull request by gitstart**
fixes https://github.com/thundernest/k-9/issues/6675
### PR Description
This PR migrates deprecated `JobIntentService` to `workManager`

### Demo v
[device-2023-06-09-131720.webm](https://github.com/GitStartHQ/client-k9-android-app/assets/35276272/6ddc0d11-73f9-4ecf-9cf6-182b9b0d9d83)

---
This code was written and reviewed by GitStart Community. Growing great engineers, one PR at a time.

### CODE REVIEW

File: app/ui/legacy/src/main/java/com/fsck/k9/account/AccountRemoverWorker.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import androidx.work.OneTimeWorkRequest
+import androidx.work.WorkManager
+import androidx.work.WorkerParameters
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import org.koin.core.component.KoinComponent
+import org.koin.core.component.inject
+
+/**
+ * An [AccountRemoverWorker] to remove an account in the background.
```
Comment (cketti) [inline app/ui/legacy/src/main/java/com/fsck/k9/account/AccountRemoverWorker.kt:15]:
```suggestion
 * A [Worker] to remove an account in the background.
```

File: app/ui/legacy/src/main/java/com/fsck/k9/account/AccountRemoverWorker.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ */
+class AccountRemoverWorker(
+    context: Context,
+    workerParams: WorkerParameters,
+) : CoroutineWorker(context, workerParams), KoinComponent {
```
Comment (cketti) [inline app/ui/legacy/src/main/java/com/fsck/k9/account/AccountRemoverWorker.kt:20]: We're not calling a suspending function in `doWork()`. So using a regular `Worker` will do.

File: app/ui/legacy/src/main/java/com/fsck/k9/account/AccountRemoverWorker.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    private val accountRemover: AccountRemover by inject()
```
Comment (cketti) [inline app/ui/legacy/src/main/java/com/fsck/k9/account/AccountRemoverWorker.kt:22]: We're using a custom `WorkerFactory` that uses Koin to instantiate workers. See `K9WorkerFactory`.

Make `accountRemover` a constructor parameter and add a `factory` entry for `AccountRemoverWorker` to an appropriate `KoinModule.kt` (see `MailSyncWorker` for an example).

File: app/ui/legacy/src/main/java/com/fsck/k9/account/AccountRemoverWorker.kt [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    companion object {
+        private const val ARG_ACCOUNT_UUID = "accountUuid"
+
+        fun enqueueRemoveAccountJob(context: Context, accountUuid: String) {
+            val data = Data.Builder()
+                .putString(ARG_ACCOUNT_UUID, accountUuid)
+                .build()
+
+            val request = OneTimeWorkRequest.Builder(AccountRemoverWorker::class.java)
```
Comment (cketti) [inline app/ui/legacy/src/main/java/com/fsck/k9/account/AccountRemoverWorker.kt:41]: This should be expedited work. See https://developer.android.com/guide/background/persistent/getting-started/define-work#expedited

File: app/ui/legacy/src/main/java/com/fsck/k9/account/AccountRemoverWorker.kt [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+        accountRemover.removeAccount(accountUuid)
+
+        Result.success()
+    }
+
+    companion object {
+        private const val ARG_ACCOUNT_UUID = "accountUuid"
+
+        fun enqueueRemoveAccountJob(context: Context, accountUuid: String) {
```
Comment (cketti) [inline app/ui/legacy/src/main/java/com/fsck/k9/account/AccountRemoverWorker.kt:36]:
```suggestion
        fun enqueueRemoveAccountWorker(context: Context, accountUuid: String) {
```

**Review by cketti — CHANGES_REQUESTED** [review]

**Comment by gitstart** [comment]
Thank you for reviewing our PR @cketti,
Requested changes are pushed. Please feel free to review it again.. 🙏

File: app/ui/legacy/src/main/java/com/fsck/k9/account/AccountRemoverWorker.kt [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    workerParams: WorkerParameters,
+) : Worker(context, workerParams) {
+
+    override fun doWork(): Result {
+        val accountUuid = inputData.getString(ARG_ACCOUNT_UUID)
+            ?: throw IllegalArgumentException("No account UUID provided")
+
+        val success = accountRemover.removeAccount(accountUuid)
+
+        return if (success) Result.success() else Result.retry()
```
Comment (cketti) [inline app/ui/legacy/src/main/java/com/fsck/k9/account/AccountRemoverWorker.kt:26]: `AccountRemover.removeAccount()` returns `false` when the account UUID couldn't be found, i.e. the account has already been removed. Retrying will lead to the same result, so this operation is retried forever.

Currently there's no situation where we want to retry the account removal operation. Please change the code to always return `Result.success()`.

**Review by cketti — CHANGES_REQUESTED** [review]

File: app/ui/legacy/src/main/java/com/fsck/k9/account/AccountRemoverWorker.kt (hunk 6 above)
Comment (cketti) [reply app/ui/legacy/src/main/java/com/fsck/k9/account/AccountRemoverWorker.kt]: Fixed in d8e09b5790147a605e4622eae6cc59a6be48e7dc

**Review by cketti — APPROVED** [review]

**Pull request merged**